### PR TITLE
Fix nil panic when overwriting metadata

### LINKED ISSUE #6381: json.overwrite_keys causes filebeat to panic
I am using one instance of filebeat to store docker logs in local log file, each line is json. Another filebeat is reading this file and sending the logs to elasticsearch. This second filebeat does not work if I enable json.overwrite_keys: true.

My config:

```
name: "myname"
path:
  home: /opt/filebeat
  data: "/app/volumes/data/logforwarder"

filebeat.prospectors:
  - type: log
    paths:
      - "/app/volumes/logs/docker-json.log*"
    json:
      keys_under_root: true
      add_error_key: true
      message_key: message
      overwrite_keys: true
    close_inactive: 1440h

fields:
  machine_group: "mg1"

setup.template.enabled: false

output.elasticsearch:
  hosts: ["eshost:9210"]
  index: "mg1-dockercontainer-%{+yyyy.MM}"
  protocol: https
  ssl:
    certificate_authorities: ["/app/volumes/config/certs/elasticsearch-ca.crt"]
    enabled: true
    verification_mode: full
    supported_protocols: [ "TLSv1.2" ]
  username: sl
  password: sl
```

Filebeat log:

```
log/harvester.go:216    Harvester started for file: /app/volumes/logs/docker-json.log
panic: assignment to entry in nil map

goroutine 73 [running]:
github.com/elastic/beats/libbeat/common.MapStr.Update(0x0, 0xc4204af0b0)
        /go/src/github.com/elastic/beats/libbeat/common/mapstr.go:43 +0xd7
github.com/elastic/beats/libbeat/common/jsontransform.WriteJSONKeys(0xc420053bb8, 0xc4204af080, 0x16ca001)
        /go/src/github.com/elastic/beats/libbeat/common/jsontransform/jsonhelper.go:50 +0xa9b
github.com/elastic/beats/filebeat/harvester/reader.MergeJSONFields(0xc4204af2c0, 0xc4204af080, 0xc420053ce0, 0x0, 0x0, 0x10101, 0xc420017260, 0x21, 0xfd65)
        /go/src/github.com/elastic/beats/filebeat/harvester/reader/json.go:120 +0x1c3
github.com/elastic/beats/filebeat/prospector/log.(*Harvester).Run(0xc42009b900, 0x0, 0x0)
        /go/src/github.com/elastic/beats/filebeat/prospector/log/harvester.go:287 +0x98a
github.com/elastic/beats/filebeat/harvester.(*Registry).Start.func1(0xc42005ccc0, 0x1f3c240, 0xc42009b900)
        /go/src/github.com/elastic/beats/filebeat/harvester/registry.go:81 +0x95
created by github.com/elastic/beats/filebeat/harvester.(*Registry).Start
        /go/src/github.com/elastic/beats/filebeat/harvester/registry.go:74 +0x148
```

Example entry from /app/volumes/logs/docker-json.log:
{"@timestamp":"2018-02-14T08:58:13.484Z","@metadata":{"beat":"filebeat","type":"doc","version":"6.2.1"},"stream":"stdout","message":"[2018-02-14T09:58:13,483][INFO ][logstash.agent           ] Pipelines running {:count=\u003e2, :pipelines=\u003e[\".monitoring-logstash\", \"beatsmain\"]}","prospector":{"type":"docker"},"docker":{"container":{"id":"48425cf7a58af112ff950ddcebe06968aed2d589bd9729422e98a4c9f7bd2a56","name":"els","image":"localhost/elk/logstash:latest","labels":{"com":{"docker":{"compose":{"version":"1.18.0","config-hash":"0993dddd4bf7d23bb20d04392b9fcb853da11d00a4b5d615ecf31af2be02f26b","container-number":"1","oneoff":"False","project":"elspo","service":"logstash"}}}}}},"beat":{"version":"6.2.1","name":"myname","hostname":"7a0006d02784"},"source":"/var/lib/docker/containers/48425cf7a58af112ff950ddcebe06968aed2d589bd9729422e98a4c9f7bd2a56/48425cf7a58af112ff950ddcebe06968aed2d589bd9729422e98a4c9f7bd2a56-json.log","offset":21400}

Affected filebeat versions: at least 6.1.x, 6.2.x

### PR DESCRIPTION
**Pull request by sincejune**
## What does this PR do?

I'm using Filebeat writing data into Kafka and having another Filebeat reading from Kafka. I got this error when I was trying to decode the `message` field into root-level via `decode_json_fields` processor.

```
panic: assignment to entry in nil map

goroutine 105 [running]:
github.com/elastic/beats/libbeat/common.MapStr.Update(0x0, 0xc0001620c0)
	/go/src/github.com/elastic/beats/libbeat/common/mapstr.go:61 +0xd3
github.com/elastic/beats/libbeat/common/jsontransform.WriteJSONKeys(0xc0000b8080, 0xc000162090, 0x101)
	/go/src/github.com/elastic/beats/libbeat/common/jsontransform/jsonhelper.go:67 +0x9ad
github.com/elastic/beats/libbeat/processors/actions.(*decodeJSONFields).Run(0xc0001710c0, 0xc0000b8080, 0xc00017bb00, 0x0, 0x0)
	/go/src/github.com/elastic/beats/libbeat/processors/actions/decode_json_fields.go:145 +0x751
github.com/elastic/beats/libbeat/publisher/processing.(*group).Run(0xc000768d50, 0xc0000b8080, 0xc0000b8080, 0x0, 0x0)
	/go/src/github.com/elastic/beats/libbeat/publisher/processing/processors.go:104 +0x94
github.com/elastic/beats/libbeat/publisher/processing.(*group).Run(0xc000768d20, 0xc0000b8080, 0x203000, 0x7f10562985b0, 0x7f1056295ff8)
	/go/src/github.com/elastic/beats/libbeat/publisher/processing/processors.go:104 +0x94
github.com/elastic/beats/libbeat/publisher/pipeline.(*client).publish(0xc000184f80, 0xe3ed9c0, 0xed7ed39c7, 0x3da5020, 0x0, 0xc00017abd0, 0x2551760, 0xc0001b8410, 0x0)
	/go/src/github.com/elastic/beats/libbeat/publisher/pipeline/client.go:89 +0x571
github.com/elastic/beats/libbeat/publisher/pipeline.(*client).Publish(0xc000184f80, 0xe3ed9c0, 0xed7ed39c7, 0x3da5020, 0x0, 0xc00017abd0, 0x2551760, 0xc0001b8410, 0x0)
	/go/src/github.com/elastic/beats/libbeat/publisher/pipeline/client.go:68 +0xc2
github.com/elastic/beats/filebeat/channel.(*outlet).OnEvent(0xc000768f60, 0xe3ed9c0, 0xed7ed39c7, 0x3da5020, 0x0, 0xc00017abd0, 0x2551760, 0xc0001b8410, 0x0, 0x0)
	/go/src/github.com/elastic/beats/filebeat/channel/outlet.go:64 +0x7e
github.com/elastic/beats/filebeat/input/kafka.(*groupHandler).ConsumeClaim(0xc000164000, 0x2a49a00, 0xc0007b6080, 0x2a3cd60, 0xc000495c80, 0xa, 0x0)
	/go/src/github.com/elastic/beats/filebeat/input/kafka/input.go:332 +0x1c5
github.com/elastic/beats/vendor/github.com/Shopify/sarama.(*consumerGroupSession).consume(0xc0007b6080, 0xc00027f380, 0xa, 0x0)
	/go/src/github.com/elastic/beats/vendor/github.com/Shopify/sarama/consumer_group.go:615 +0x27f
github.com/elastic/beats/vendor/github.com/Shopify/sarama.newConsumerGroupSession.func2(0xc0007b6080, 0xc00027f380, 0xa, 0xc000000000)
	/go/src/github.com/elastic/beats/vendor/github.com/Shopify/sarama/consumer_group.go:544 +0xd0
created by github.com/elastic/beats/vendor/github.com/Shopify/sarama.newConsumerGroupSession
	/go/src/github.com/elastic/beats/vendor/github.com/Shopify/sarama/consumer_group.go:536 +0x526
```

## Why is it important?

The Filebeat process will crash without the fix.

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

using unittest.

## Related issues

- Closes #6381

## Note

Filebeat will actually output JSONs with duplicated keys when I'm trying to overwrite the `@metadata` field. And I'm not pretty sure if that's good for beats. Check the example below:

#### Filebeat configuration
```
filebeat.inputs:
  - type: stdin
    enabled: true
output.console:
    enabled: true

processors:
  - decode_json_fields:
      fields: [ "message" ]
      target: ""
      overwrite_keys: true
```

#### Input
```
{"@timestamp":"2021-03-24T12:17:54.978Z","@metadata":{"beat":"filebeat","type":"_doc","version":"7.0.0"},"message":"overwrite metadata test"}
```
#### Output
```
{"@timestamp":"2021-03-24T12:17:54.978Z","@metadata":{"beat":"filebeat","type":"_doc","version":"8.0.0","type":"_doc","version":"7.0.0","beat":"filebeat"},"log":{"offset":0,"file":{"path":""}},"message":"overwrite metadata test","input":{"type":"stdin"},"agent":{"name":"MacBook","type":"filebeat","version":"8.0.0","ephemeral_id":"ephemeral_id","id":"id"},"ecs":{"version":"1.8.0"},"host":{"name":"MacBook"}}
```

### CODE REVIEW

File: libbeat/common/jsontransform/jsonhelper.go [hunk 1]
```diff
@@ -77,6 +77,10 @@ func WriteJSONKeys(event *beat.Event, keys map[string]interface{}, expandKeys, o
 			event.Timestamp = ts
 
 		case "@metadata":
+			if event.Meta == nil {
+				event.Meta = common.MapStr{}
+			}
+
 			switch m := v.(type) {
 			case map[string]string:
```
Comment (jsoriano) [inline libbeat/common/jsontransform/jsonhelper.go:85]: Is this needed in all cases or only in the `map[string]string`  case? I wonder if this should be left as `nil` in other cases.

```suggestion
			switch m := v.(type) {
			case map[string]string:
				if event.Meta == nil && len(m) > 0 {
					event.Meta = common.MapStr{}
				}
```

Comment (sincejune) [reply libbeat/common/jsontransform/jsonhelper.go]: Actually, my real case hit the `map[string]interface{}` case.

I modified the code only to update `event.Meta` in these two cases.

ref: https://github.com/elastic/beats/pull/24741/commits/c95f2c66303d0396c6a7ed28258736146d228437

Comment (jsoriano) [reply libbeat/common/jsontransform/jsonhelper.go]: Thanks!

**Review by jsoriano — APPROVED** [review]
Waiting for green, thanks!

**Comment by jsoriano** [comment]
@sincejune thanks a lot for this fix! I merged it and it will be available in 7.13.

**Comment by sincejune** [comment]
@jsoriano You're welcome. Thanks for taking care of this!

**Pull request merged**